compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaBdChain.cpp

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/aie_runtime/Utils/ChannelBdIdGenerator.h"
+#include "iree-amd-aie/aie_runtime/iree_aie_runtime.h"
+#include "mlir/IR/Iterators.h"
+
+#define DEBUG_TYPE "iree-amdaie-dma-bd-chain"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+LogicalResult dmaBdChain(AMDAIE::WorkgroupOp workgroupOp) {
+  IRRewriter rewriter(workgroupOp->getContext());
+
+  DenseMap<std::pair<uint32_t, uint32_t>, uint32_t> colArgIdxToNextBdId;
+  DenseMap<std::pair<uint32_t, uint32_t>, uint32_t> colArgIdxToStartBdId;
+  bool useNextBd = false;
+  bool enablePacket = false;
+  uint32_t argIdx = 0;
+  uint32_t col = 0;
+  uint32_t repeatCount = 1;
+
+  // Walk the operations in the reverse order to build the chain
+  AMDAIE::ControlCodeOp controlCodeOp = workgroupOp.getControlCode();
+  WalkResult res = controlCodeOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](Operation *op) {
+        if (auto npuPushToQueueOp = dyn_cast<AMDAIE::NpuPushToQueueOp>(op)) {
+          // repeat count > 1, do not chain BDs
+          repeatCount = npuPushToQueueOp.getRepeatCount();
```

Comment:
Yeah I agree, working on that

Reply:
Done, though one thing is with the latest `main` branch, the performance tests all have `repeat_count > 1`,  making this PR "zero grain".... Not sure why `next_bd` and `repeat_count`  doesn't work together. My guess is having `repeat_count` on a chain means go through the whole chain then repeat from the start of the chain, instead of first repeating on individual BD multiple times and then moving to the next BD on the chain?

Reply:
Right, it's likely this:

```
Example: 3 BD chained with “next-BD”: A → B → C.
Task: Start-BD-ID: A; Repeat-count: 3
Sequence of BDs executed: A, B, C, A, B, C, A, B, C <token>
```

Reply:
You could maybe try first without chaining, as the shim channels have a task queue of size `4`, so that could be an implicit chain of 4 BDs (with different bd IDs) before waiting (each one needs a push_to_queue).

Reply:
Just double checking, for queue size of `4`, it means 4 BDs or 4 chains, since each chain only has one `push_to_queue`?

Reply:
Yeah, 4 `push_to_queue`'s /chains